titan/utils.py

```diff
@@ -333,3 +336,129 @@ def set_up_logging(params):
             format=log_msg_format,
             level=params.outputs.logging.level,
         )
+
+
+# if edge exists, add it to edges - sorted by name to avoid duplicates as edges are undirected
+def add_edge(edges, loc, edge):
+    if edge != "" and edge != loc:
+        edges.add(tuple(sorted([loc, edge])))
+
+
+def grid_file_to_edges(file_path: str, diagonal_neighbors: bool = False) -> Dict:
+    """
+    Read a csv describing the layout of locations and return a dictionary describing the location edges, which can then be used in the params [location.edges].
+
+    Sample csv:
+    ```
+    location_1,location_3,location_4
+    location_1,location_3,location_4
+    location_1,location_3,location_4
+    location_1,location_2,
+    location_2,location_2,
```

Comment:
Why repeat the same line three times? For the diagonal neighbors thing, but wouldn't you just need it twice for that?

Reply:
need, no.  But that's just the geography I made up in my head?   Is it confusing?